pyro/contrib/forecast/forecaster.py

```diff
@@ -0,0 +1,220 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from abc import ABCMeta, abstractmethod
+
+import torch
+import torch.nn as nn
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.infer import SVI, Trace_ELBO
+from pyro.infer.autoguide import AutoNormal, init_to_sample
+from pyro.nn.module import PyroModule
+from pyro.optim import ClippedAdam
+
+from .util import PrefixConditionMessenger, PrefixReplayMessenger, reshape_batch
+
+logger = logging.getLogger(__name__)
+
+
+class _ForecastingModelMeta(type(PyroModule), ABCMeta):
+    pass
+
+
+class ForecastingModel(PyroModule, metaclass=_ForecastingModelMeta):
+    """
+    Abstract base class for forecasting models.
+
+    Derived classes must implement the :meth:`model` method.
+    """
+    def __init__(self):
+        super().__init__()
+        self._prefix_condition_data = {}
+
+    @abstractmethod
+    def model(self, zero_data, covariates):
+        """
+        Generative model definition.
+
+        Implementations must call the :meth:`predict` method exactly once.
+
+        Implementations must draw all time-dependent noise inside the
+        :meth:`time_plate`. The prediction passed to :meth:`predict` must be a
+        deterministic function of noise tensors that are independent over time.
+        This requirement is slightly more general than state space models.
+
+        :param zero_data: A zero tensor like the input data, but extended to
+            the duration of the :meth:`time_plate`. This allows models to
+            depend on the shape and device of data but not its value.
+        :type zero_data: ~torch.Tensor
+        :param covariates: A tensor of covariates with time dimension -2.
+        :type covariates: ~torch.Tensor
+        :returns: Return value is ignored.
+        """
+        raise NotImplementedError
+
+    @property
+    def time_plate(self):
+        """
+        :returns: A plate named "time" with size ``covariates.size(-2)`` and
+            ``dim=-1``. This is available only during model execution.
+        :rtype: :class:`~pyro.plate`
+        """
+        assert self._time_plate is not None, ".time_plate accessed outside of .model()"
+        return self._time_plate
+
+    def predict(self, noise_dist, prediction):
```

Comment:
should there be a comment that this is analogous to an observe statement in a typical pyro model?

Reply:
done